// @elog/core@0.5.0-beta.9/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } } function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _createStarExport(obj) { Object.keys(obj) .filter((key) => key !== "default" && key !== "__esModule") .forEach((key) => { if (exports.hasOwnProperty(key)) { return; } Object.defineProperty(exports, key, {enumerable: true, configurable: true, get: () => obj[key]}); }); }var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined")
    return require.apply(this, arguments);
  throw new Error('Dynamic require of "' + x + '" is not supported');
});
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/client.ts
var _sdkyuque = require('@elog/sdk-yuque'); var _sdkyuque2 = _interopRequireDefault(_sdkyuque); _createStarExport(_sdkyuque);
var _sdknotion = require('@elog/sdk-notion'); var _sdknotion2 = _interopRequireDefault(_sdknotion); _createStarExport(_sdknotion);
var _sdkflowus = require('@elog/sdk-flowus'); var _sdkflowus2 = _interopRequireDefault(_sdkflowus); _createStarExport(_sdkflowus);
var _deploy = require('@elog/deploy'); var _deploy2 = _interopRequireDefault(_deploy); _createStarExport(_deploy);
var _pluginimage = require('@elog/plugin-image'); var _pluginimage2 = _interopRequireDefault(_pluginimage); _createStarExport(_pluginimage);
var _shared = require('@elog/shared'); _createStarExport(_shared);
var _fs = require('fs'); var fs = _interopRequireWildcard(_fs);
var _path = require('path'); var path = _interopRequireWildcard(_path);
var _process = require('process'); var process = _interopRequireWildcard(_process);
var Elog = class {
  constructor(config) {
    /** 缓存文章 */
    this.cachedArticles = [];
    /** 是否需要更新，当所有文章都不需要更新，这个标记就会阻止后续流程 */
    this.needUpdate = false;
    /** 待更新的文章列表 */
    this.needUpdateArticles = [];
    this.config = config;
    this.initIncrementalUpdate(config);
    this.initWritingPlatform(config);
    this.initDeployPlatform(config);
    this.initImgCdn(config);
  }
  /**
   * 初始化增量配置
   * @param config
   */
  initIncrementalUpdate(config) {
    try {
      const cacheJson = __require(path.join(process.cwd(), config.extension.cachePath));
      const { docs } = cacheJson;
      this.cachedArticles = docs || [];
    } catch (error) {
      _shared.out.access("\u5168\u91CF\u66F4\u65B0", "\u672A\u83B7\u53D6\u5230\u7F13\u5B58\uFF0C\u5C06\u5168\u91CF\u66F4\u65B0\u6587\u6863");
    }
  }
  /**
   * 初始化写作平台
   * @param config
   */
  initWritingPlatform(config) {
    if (config.write.platform === "yuque" /* YUQUE */) {
      let yuqueConfig = config.write.yuque;
      this.downloaderClient = new (0, _sdkyuque2.default)(yuqueConfig);
    } else if (config.write.platform === "notion" /* NOTION */) {
      let notionConfig = config.write.notion;
      this.downloaderClient = new (0, _sdknotion2.default)(notionConfig);
    } else if (config.write.platform === "flowus" /* FLOWUS */) {
      let flowusConfig = config.write.flowus;
      this.downloaderClient = new (0, _sdkflowus2.default)(flowusConfig);
    }
  }
  /**
   * 初始化部署平台
   * @param config
   */
  initDeployPlatform(config) {
    const deployOptions = config.deploy;
    this.deployClient = new (0, _deploy2.default)(deployOptions);
  }
  /**
   * 初始化图片转CDN配置
   * @param config
   */
  initImgCdn(config) {
    var _a;
    if ((_a = config.image) == null ? void 0 : _a.enable) {
      if (config.write.platform === "flowus" /* FLOWUS */) {
        process.env.REFERER_URL = "https://flowus.cn/";
      }
      this.imageClient = new (0, _pluginimage2.default)(config.image);
    }
  }
  /**
   * 下载文章详情列表
   */
  fetchArticles() {
    return __async(this, null, function* () {
      var _a;
      let articleList = yield this.downloaderClient.getDocList();
      if (!(articleList == null ? void 0 : articleList.length)) {
        this.needUpdate = false;
        return;
      }
      this.cachedArticles = this.cachedArticles.filter(
        (cache) => articleList.findIndex((item) => item.doc_id === cache.doc_id) !== -1
      );
      let ids = [];
      let idMap = {};
      for (const article of articleList) {
        const cacheIndex = this.cachedArticles.findIndex(
          (cacheItem) => cacheItem.doc_id === article.doc_id
        );
        if (cacheIndex < 0) {
          ids.push(article.doc_id);
          idMap[article.doc_id] = {
            status: "create" /* create */
          };
        } else {
          const cacheArticle = this.cachedArticles[cacheIndex];
          const cacheAvailable = article.updated === cacheArticle.updated;
          if (!cacheAvailable) {
            ids.push(article.doc_id);
            idMap[article.doc_id] = {
              index: cacheIndex,
              status: "update" /* update */
            };
          }
        }
      }
      if (!ids.length) {
        this.needUpdate = false;
        return;
      }
      this.needUpdate = true;
      let docDetailList = yield this.downloaderClient.getDocDetailList(ids);
      if ((_a = this.config.image) == null ? void 0 : _a.enable) {
        docDetailList = yield this.processImage(docDetailList);
      }
      this.needUpdateArticles = docDetailList;
      for (const docDetail of docDetailList) {
        const { index, status } = idMap[docDetail.doc_id];
        if (status === "create" /* create */) {
          this.cachedArticles.push(docDetail);
        } else {
          this.cachedArticles[index] = docDetail;
        }
      }
    });
  }
  /**
   * 写入语雀的文章缓存 json 文件
   */
  writeArticleCache() {
    var _a;
    try {
      let catalog = [];
      if (this.config.write.platform === "yuque" /* YUQUE */) {
        const yuqueClient = this.downloaderClient;
        catalog = yuqueClient.ctx.catalog;
      } else if (this.config.write.platform === "notion" /* NOTION */) {
        const notionClient = this.downloaderClient;
        catalog = notionClient.ctx.catalog;
      } else if (this.config.write.platform === "flowus" /* FLOWUS */) {
        const flowusClient = this.downloaderClient;
        catalog = flowusClient.ctx.catalog;
      }
      let cacheDocs = this.cachedArticles.map((item) => {
        return {
          id: item.id,
          doc_id: item.doc_id,
          title: item.doc_id,
          updated: item.updated,
          body_original: item.body_original,
          properties: item.properties,
          catalog: item.catalog,
          body: ""
        };
      });
      if ((_a = this.config.extension) == null ? void 0 : _a.isFullCache) {
        cacheDocs = this.cachedArticles;
      }
      const cacheJson = {
        docs: cacheDocs,
        catalog
      };
      fs.writeFileSync(this.config.extension.cachePath, JSON.stringify(cacheJson, null, 2), {
        encoding: "utf8"
      });
    } catch (e) {
      _shared.out.warning("\u7F13\u5B58\u5931\u8D25", `\u5199\u5165\u7F13\u5B58\u4FE1\u606F\u5931\u8D25\uFF0C\u8BF7\u68C0\u67E5${e.message}`);
    }
  }
  /**
   * 处理文章图片
   */
  processImage(docDetailList) {
    return __async(this, null, function* () {
      return yield this.imageClient.replaceImages(docDetailList);
    });
  }
  /**
   * 部署文章
   */
  deployArticles() {
    return __async(this, null, function* () {
      yield this.deployClient.deploy(this.needUpdateArticles);
    });
  }
  // 下载文档 => 增量更新文章到缓存 json 文件
  deploy() {
    return __async(this, null, function* () {
      yield this.fetchArticles();
      if (!this.needUpdate) {
        _shared.out.access("\u4EFB\u52A1\u7ED3\u675F", "\u6CA1\u6709\u9700\u8981\u66F4\u65B0\u7684\u6587\u6863");
        return;
      }
      this.writeArticleCache();
      yield this.deployArticles();
    });
  }
};
var client_default = Elog;

// src/index.ts






var src_default = client_default;


exports.default = src_default;
